File: frontend/src/types/auth02.ts
/** Domain types for auth / AuthRecord02 */
export type AuthRecord02Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface AuthRecord02 {
  id: string;
  externalCode: string;
  displayName: string;
  status: AuthRecord02Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface AuthRecord02CreateRequest {
  externalCode: string;
  displayName: string;
  status: AuthRecord02Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface AuthRecord02ListResponse {
  items: AuthRecord02[];
  total?: number;
}

export interface AuthRecord02StatsResponse {
  activeCount: number;
}

export function isAuthRecord02Active(rec: AuthRecord02): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatAuthRecord02Label(rec: AuthRecord02): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareAuthRecord02ByPriority(a: AuthRecord02, b: AuthRecord02): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}

File: backend/internal/service/auth_02.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// AuthRecord02Service encapsulates business rules for AuthRecord02.
type AuthRecord02Service struct {
	repo *repository.AuthRecord02Repository
}

// NewAuthRecord02Service wires the service.
func NewAuthRecord02Service(repo *repository.AuthRecord02Repository) *AuthRecord02Service {
	return &AuthRecord02Service{repo: repo}
}

// Create validates and persists a new AuthRecord02.
func (s *AuthRecord02Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.AuthRecord02, error) {
	now := time.Now().UTC()
	rec := &models.AuthRecord02{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a AuthRecord02 by id.
func (s *AuthRecord02Service) Get(ctx context.Context, id string) (*models.AuthRecord02, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped AuthRecord02 pages.
func (s *AuthRecord02Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.AuthRecord02, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing AuthRecord02.
func (s *AuthRecord02Service) Update(ctx context.Context, actor string, rec *models.AuthRecord02) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.UpdatedBy = actor
	rec.UpdatedAt = time.Now().UTC()
	// repository Update applies optimistic locking and increments version
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a AuthRecord02.
func (s *AuthRecord02Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds AuthRecord02 by external code prefix.
func (s *AuthRecord02Service) Search(ctx context.Context, facilityID, prefix string) ([]models.AuthRecord02, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *AuthRecord02Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}
